src/mumble/PipeWire.cpp

```diff
@@ -0,0 +1,437 @@
+void PipeWireEngine::start() {
+	if (m_thread) {
+		pws->pw_thread_loop_start(m_thread);
+	}
+}
+
+void PipeWireEngine::stop() {
+	if (m_loop) {
+		pws->pw_thread_loop_stop(m_thread);
+	}
+}
+
+pw_buffer *PipeWireEngine::dequeueBuffer() {
+	if (m_stream) {
+		return pws->pw_stream_dequeue_buffer(m_stream);
+	}
+
+	return nullptr;
+}
+
+void PipeWireEngine::queueBuffer(pw_buffer *buffer) {
+	if (m_stream) {
+		pws->pw_stream_queue_buffer(m_stream, buffer);
+	}
+}
+
+PipeWireInput::PipeWireInput() {
+	m_engine = std::make_unique< PipeWireEngine >("Capture", processCallback);
+	if (!m_engine->isOk()) {
+		return;
+	}
+
+	iMicChannels = Global::get().s.qsPipeWireInput.toUShort();
+
+	constexpr uint32_t channels[]{
+		SPEAKER_FRONT_LEFT,
+		SPEAKER_FRONT_RIGHT,
+	};
+
+	if (!m_engine->connect(PW_DIRECTION_INPUT, channels, iMicChannels)) {
+		return;
+	}
+
+	eMicFormat = SampleFloat;
+	iMicFreq   = SAMPLE_RATE;
+	initializeMixer();
+
+	m_engine->start();
+}
+
+PipeWireInput::~PipeWireInput() {
+	m_engine->stop();
+}
+
+void PipeWireInput::processCallback(void *param) {
+	auto pwi = static_cast< PipeWireInput * >(Global::get().ai.get());
+	if (!pwi) {
+		return;
+	}
+
+	auto engine = static_cast< PipeWireEngine * >(param);
+
+	auto buffer = engine->dequeueBuffer();
+	if (!buffer) {
+		return;
+	}
+
+	auto data = buffer->buffer->datas[0];
+	if (!data.data) {
+		return;
+	}
+
+	pwi->addMic(data.data, data.chunk->size / sizeof(float));
+
+	engine->queueBuffer(buffer);
+}
+
+void PipeWireInput::run() {
+}
+
+PipeWireOutput::PipeWireOutput() {
+	m_engine = std::make_unique< PipeWireEngine >("Playback", processCallback);
+	if (!m_engine->isOk()) {
+		return;
+	}
+
+	iChannels = Global::get().s.qsPipeWireOutput.toUShort();
+
+	// This mapping works for:
+	//
+	// - Mono
+	// - Stereo
+	// - 2.1 surround
+	// - 3.1 surround
+	// - 5.1 surround
+	// - 7.1 surround
+	//
+	// Ideally this should be configurable by the user.
+	constexpr uint32_t channels[]{ SPEAKER_FRONT_LEFT, SPEAKER_FRONT_RIGHT, SPEAKER_LOW_FREQUENCY, SPEAKER_FRONT_CENTER,
+								   SPEAKER_BACK_LEFT,  SPEAKER_BACK_RIGHT,  SPEAKER_SIDE_LEFT,     SPEAKER_SIDE_RIGHT };
+
+	if (!m_engine->connect(PW_DIRECTION_OUTPUT, channels, iChannels)) {
+		return;
+	}
+
+	eSampleFormat = SampleFloat;
+	iMixerFreq    = SAMPLE_RATE;
+	initializeMixer(channels);
+
+	m_engine->start();
+}
+
+PipeWireOutput::~PipeWireOutput() {
+	m_engine->stop();
+}
+
+void PipeWireOutput::processCallback(void *param) {
+	auto pwo = static_cast< PipeWireOutput * >(Global::get().ao.get());
+	if (!pwo) {
+		return;
+	}
+
+	auto engine = static_cast< PipeWireEngine * >(param);
+
+	auto buffer = engine->dequeueBuffer();
+	if (!buffer) {
+		return;
+	}
+
+	auto data = buffer->buffer->datas[0];
+	if (!data.data) {
+		return;
+	}
+
+	auto chunk = data.chunk;
+	if (!chunk) {
+		return;
+	}
+
+	chunk->offset = 0;
+	chunk->stride = sizeof(float) * pwo->iChannels;
+
+	const auto frames = qMin(data.maxsize / chunk->stride, pwo->iFrameSize);
```

Comment:
`auto` hides type

Also: `qMin` should not be used anymore (afaik). Instead `std::min` is preferred